tensorboard/webapp/metrics/store/metrics_reducers_test.ts

```diff
@@ -2500,6 +2510,189 @@ describe('metrics reducers', () => {
         const state3 = reducers(state2, actions.selectTimeEnableToggled());
         expect(state3.selectTimeEnabled).toBe(false);
       });
+
+      it('sets cardStepIndex to step 0 when selectedTime is null before toggling', () => {
+        const state1 = buildMetricsState({
+          selectTimeEnabled: false,
+          cardMetadataMap,
+          stepMinMax: {min: Infinity, max: -Infinity},
+          timeSeriesData: {
+            ...buildTimeSeriesData(),
+            images: {
+              tagC: {
+                111: {
+                  runToLoadState: {},
+                  runToSeries: {
+                    'test run Id': [
+                      {step: 0, wallTime: 0, imageId: '1'},
+                      {step: 10, wallTime: 0, imageId: '1'},
+                      {step: 20, wallTime: 10, imageId: '2'},
+                      {step: 30, wallTime: 15, imageId: '3'},
+                    ],
+                  },
+                },
+              },
+            },
+          },
+          cardStepIndex: {[imageCardId]: 2},
+        });
+
+        const state2 = reducers(state1, actions.selectTimeEnableToggled());
+
+        expect(state2.cardStepIndex).toEqual({[imageCardId]: 0});
+      });
+
+      it('sets cardStepIndex to min step when selectedTime is null before toggling', () => {
```

Comment:
I'm even more confident that you don't need this test case. Its definitely covered by `sets cardStepIndex to step 0 when selectedTime is null before toggling` and `sets selectedTime to min step when selectedTime is null before toggling`

Reply:
This way it does feel like the whole direction for test cardStepIndex is tweak toward test selectedTime.
I think there are two groups:
1.
sets cardStepIndex to step 0
sets cardStepIndex to min step
2.
sets selectedTime to step 0
sets selectedTime to min step

1-1 and 2-1 should happen together, same for 1-2 and 2-2
but somehow you pick 1-1 and 2-2. Which cause me confused and feel unstructured.

I also thought about combining them to 
sets cardStepIndex and selectedTime to step 0
sets cardStepIndex and selectedTime to min step

which can result longer description but would that satisfy your concern of repeated tests?

Reply:
I don't think your cardStepIndex tests have to care about either of these edge cases ( "step is 0", "step is min") - it doesn't really have anything to do with cardStepIndex. 

This is my methodology:

I start by looking at the code like I'm a code coverage tool: I look at all the code in the reducer and make sure that each line of code is covered by at least one test. I make sure that each branch/decision (if statements evaluating true and false, trinary operators evaluating true and false, for loops for multiple elements, etc...) is covered by a test.

When I do that, these are the test cases I see:

* selectedTime is not reinitialized if it doesn't have to be
* selectedTime is initialized with start step 0 when appropriate
* selectedTime is initialized with start step min when appropriate
* we call generateNextCardStepIndexFromSelectedTime when selectTimeEnabled and do something with the result
* we don't call generateNextCardStepIndexFromSelectedTime when !selectTimeEnabled (and so you effectively ignore the result it would return if you did call it)

It's not a perfect mechanism but it usually identifies 80 or 90% of the tests you want to write. To supplement, I also look for places where there might be risk of subtle bugs: Using the wrong variable in place of another, for example. When I do that, this is what I see:

* we pass the new  selectedTime value to generateNextCardStepIndexFromSelectedTime instead of the old null one (nextSelectedTime instead of state.selectedTime). 

What I don't see when I do this sort of analysis is:
* cardStepIndex cares about whether selectedTime.start.step = 0
* cardStepIndex cares about whether selectedTime.start.min = 0

Both those cases are just side-effects of initializing selectedTime and have nothing to do with cardStepIndex in my mind.

To think about it yet another way: If you modified the lines that initialize selectedTime step to 0 or min, do you have at least one test that fails? If you already have one, do you really need another?

Reply:
Thanks for the comment.
Updated `cardStepIndex` related test to only test "selectedTime.start.step = 0".